src/gutil_idlepool.c

```diff
@@ -148,34 +148,36 @@ void
 gutil_idle_pool_destroy(
     GUtilIdlePool* self) /* Since 1.0.34 */
 {
-    gutil_idle_pool_drain(self);
-    gutil_idle_pool_unref(self);
+    if (G_LIKELY(self)) {
+        gutil_idle_pool_drain(self);
+        gutil_idle_pool_unref(self);
+    }
 }
 
 void
 gutil_idle_pool_drain(
     GUtilIdlePool* self)
 {
-    if (G_LIKELY(self)) {
+    /* If no pool is specified, use the default one for this thread */
+    if (!self) {
+        self = gutil_idle_pool_get_default();
+    }
+    while (self->first) {
```

Comment:
I don't quite follow how this loop works, it sets self->first to NULL inside the loop. Is the something I'm missing?

Reply:
During the loop, the `destroy` callbacks get invoked which can do whatever e.g. push another item to the same pool, in which case the drain loop will be repeated. That rarely happens in practice and may even cause an infinite loop (a `destroy` callback pushing a new item to the pool with a `destroy` callback pushing another item and so on) but is quite possible if you consider e.g. the `destroy` callback pointing to `g_object_unref` which happens to finalize a `GObject`. That's the thinking behind this `while` loop.